osutil/syncdir.go

```diff
@@ -63,16 +63,28 @@ var ErrSameState = fmt.Errorf("file state has not changed")
 // exhausted.
 //
 // In all cases, the function returns the first error it has encountered.
-func EnsureDirState(dir, glob string, content map[string]*FileState) (changed, removed []string, err error) {
-	if _, err := filepath.Match(glob, "foo"); err != nil {
-		panic(fmt.Sprintf("EnsureDirState got invalid pattern %q: %s", glob, err))
+func EnsureDirStateMany(dir string, globs []string, content map[string]*FileState) (changed, removed []string, err error) {
+	for _, glob := range globs {
+		if _, err := filepath.Match(glob, "foo"); err != nil {
+			panic(fmt.Sprintf("EnsureDirState got invalid pattern %q: %s", glob, err))
+		}
 	}
 	for baseName := range content {
 		if filepath.Base(baseName) != baseName {
 			panic(fmt.Sprintf("EnsureDirState got filename %q which has a path component", baseName))
 		}
-		if ok, _ := filepath.Match(glob, baseName); !ok {
-			panic(fmt.Sprintf("EnsureDirState got filename %q which doesn't match the glob pattern %q", baseName, glob))
+		sane := false
+		for _, glob := range globs {
+			if ok, _ := filepath.Match(glob, baseName); ok {
+				sane = true
+				break
+			}
+		}
+		if !sane {
+			if len(globs) == 1 {
+				panic(fmt.Sprintf("EnsureDirState got filename %q which doesn't match the glob pattern %q", baseName, globs[0]))
```

Comment:
Same as above for the panic and prefix, and also no need to special case the 1 case for an internal error. The general message looks fine.